tools/sigclient/pkg/query/functional.go

```diff
@@ -0,0 +1,567 @@
+// Copyright (c) 2021-2024 SigScalr, Inc.
+//
+// This file is part of SigLens Observability Solution
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package query
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+	"verifier/pkg/utils"
+
+	"github.com/fasthttp/websocket"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type Hits struct {
+	TotalMatched interface{}
+	Records      []map[string]interface{} // key: column name, value: column value
+}
+
+type BucketHolder struct {
+	GroupByValues []string
+	MeasureVal    map[string]interface{} // key: measure name, value: measure value
+}
+
+type Result struct {
+	TotalMatched     interface{}
+	Records          []map[string]interface{}
+	AllColumns       []string
+	ColumnsOrder     []string
+	GroupByCols      []string
+	MeasureFunctions []string
+	MeasureResults   []BucketHolder
+	Qtype            string
+	BucketCount      int
+}
+
+func CreateListOfMap(resp interface{}) []map[string]interface{} {
+	items, isList := resp.([]interface{})
+	if !isList {
+		log.Fatalf("CreateListOfMap: Records is not a list")
```

Comment:
nit, `....resp is not a list, received type: %T", resp)`

Reply:
addressed in the latest commit.